### ISSUE TITLE
Support for pre-built mod

### ISSUE THREAD

**Issue by namkazt** [JHawkley/Harmony-Injector#1]
I have this thing on my mind on every long time and not have a chance to do it.  Any ideas?

pre-built mod: 
- mod.info
- mod.dll

It probably speeds up the loading process at the start and avoids unnecessary compile and error

p/s: I'm not sure if it will be working on another os than windows

**Comment by JHawkley** [comment]
I'm actually working on a big wishlist document for modding in general that I'd like to present to the devs.  This proposal is included in it.  I'll throw it up on a gist once I've finished up a few of the final sections.

I have a preference toward convincing the devs to give modding some TLC rather than just trying to enhance it ourselves with patches.  I mean, yeah, we *could* do that, but then everyone who wants that change has to become dependent on Harmony Injector.

But, for the relevant section, I currently got this:

The solution that many other games have gone with is to just permit mods to package an assembly DLL that the game will load for the mod.  This essentially allows modders to build their mod using whatever toolsets and CIL-compatible languages they wish, removing the responsibility of code-compilation of the mod from the game.

While this does make it harder to inspect the code before approving the mod, this hasn't really led to many security concerns when it was utilized by other games, like RimWorld.  It also makes developing large-scale mods a little nicer since a full IDE that organizes and compiles such a project can manage the project, its references, and the build process.

This would have a few additional challenges to managing the mods, since searching for types referenced in blueprints would need to be a little more complex, but I don't think it would be too rough of a change, especially if some of the suggestions in the "Isolating Types between Mods" section are utilized.

A mod's `config.json` can configure an `assembly: string` property, which indicates the relative path to the main assembly the mod provides and that the game should load.  If the `assembly` property is defined in the mod's `config.json`, runtime compilation of C# files is then skipped for this mod altogether.

You could optionally also have a `references: bool` property that tells Caves of Qud to tell the Mono Runtime to search the mod directory for additional assemblies that the mod will require using the `AppDomain.AssemblyResolve` event before loading the mod's assembly.

**Comment by JHawkley** [comment]
Here is [the current draft of the document](https://gist.github.com/JHawkley/dfaeeb596fa5835dc37427666481f4f5) I was talking about.  Let me know if there is anything you think needs to be added of changed.  It'd be nice to get some comments on it before I drop it on them.

**Comment by namkazt** [comment]
that is a very cool document but I'm not sure if the dev willing to implement it or not 

I read something about create other `AppDomain` and load assembly into this allow we reload assembly. This is the way other app using the plugins system. But I'm not an expert on C# so maybe you know something about it?

**Comment by JHawkley** [comment]
I had a section proposing that, actually, but I removed it...

A modding system built around separated AppDomains adds a ton of complexity and would make certain mods, like Harmony Injector, impossible; cross-domain interactions have to occur through specific, well-defined interfaces.

Just getting vanilla modding working again would be an enormous challenge; most everything about how they currently manage objects in blueprints would need to be fundamentally altered.  The way they load the game's configuration now, using reflection, would not be possible with separated AppDomains.

So, I removed it because it seems like an impractical investment; giving up so much flexibility, adding tons of complexity, all just so that the game doesn't need to be restarted on the rare occasion that the list of mods-to-be-loaded changes while the game is running.

Besides, if you fear they're not willing to implement something as simple as "restart the game if the mods change..."  There is no hope for AppDomains.  😝

**Comment by namkazt** [comment]
So I will test the current version of Harmony Injector and submit it. Are you able to update it?

**Comment by JHawkley** [comment]
I'm not sure what you mean.  Update it?  Are you asking if I'm still around to help you check to make sure the Workshop submission updates and appears to function correctly?

If that is what you meant, then yeah.  I can help with that.

**Comment by namkazt** [comment]
:D yeah, I saw some mod projects with multi contributors but I'm not sure how. 😝

**Comment by JHawkley** [comment]
Ahh...  So you're wondering if *I* can submit it to Steam in place of yourself.  Now, that I didn't try.

Anyways, relevant to this issue, I started exploring the idea of loading pre-compiled mods.  I thought it was out of scope of Harmony Injector, so I started a new mod for it.

The idea was to add some attributes to Harmony Injector that allows other mods to mark methods that should be called at key points during the injection process.

With this, I could create a mod that could step in just before the call to `BuildScriptMods`, patch in some compatibility for loading mod assemblies (like make sure the Mod Approval system recognizes them as script mods), then after the build, do the loading of the assemblies.

Well...  I stepped on another compiler bug; custom attributes cause it to shit a brick.

```
<color=#AAAAAA>[Info] Compiling mod assembly...</color>
<color=orange>XRL.ModManager.BuildScriptMods</color>
<color=red></color>

Crash!!!
SymInit: Symbol-SearchPath: '.;C:\Program Files (x86)\Steam\steamapps\common\Caves of Qud;C:\Program Files (x86)\Steam\steamapps\common\Caves of Qud
OS-Version: 10.0.0

========== OUTPUTTING STACK TRACE ==================

0x08175112 (mono) mono_get_inflated_method
0x08178B88 (mono) mono_class_from_typeref
0x081CE647 (mono) mono_reflection_get_custom_attrs_blob
0x081D2EA2 (mono) mono_custom_attrs_get_attr
0x296B6AC7 (Mono JIT Code) (wrapper managed-to-native) System.Reflection.Emit.TypeBuilder:create_runtime_class (System.Reflection.Emit.TypeBuilder)
0x296B65AB (Mono JIT Code) System.Reflection.Emit.TypeBuilder:CreateType ()
0x29757CEE (Mono JIT Code) Mono.CSharp.TypeDefinition:CloseContainer ()
0x29757BD2 (Mono JIT Code) Mono.CSharp.TypeContainer:CloseContainer ()
0x29757D6C (Mono JIT Code) Mono.CSharp.TypeDefinition:CloseContainer ()
0x29757BD2 (Mono JIT Code) Mono.CSharp.TypeContainer:CloseContainer ()
0x29757BD2 (Mono JIT Code) Mono.CSharp.TypeContainer:CloseContainer ()
0x29757BD2 (Mono JIT Code) Mono.CSharp.TypeContainer:CloseContainer ()
0x29757B5F (Mono JIT Code) Mono.CSharp.ModuleContainer:CloseContainer ()
0x264FDD90 (Mono JIT Code) CSharpCompiler.CustomDynamicDriver:Compile (System.Reflection.Emit.AssemblyBuilder&,System.AppDomain,bool)
0x264FB148 (Mono JIT Code) CSharpCompiler.CodeCompiler:CompileFromCompilerSettings (Mono.CSharp.CompilerSettings,bool)
0x264FA859 (Mono JIT Code) CSharpCompiler.CodeCompiler:CompileAssemblyFromSourceBatch (System.CodeDom.Compiler.CompilerParameters,string[])
0x08CC988B (Mono JIT Code) XRL.ModManager:BuildScriptMods ()
```

Looks like it blew up as it was trying to resolve the custom attributes.  Another one to add to the list, I guess...  *SIIGHHH...*

Hmm...  Maybe I could just use an abstract class with hooks for these instead.  Just locate all classes in the mod assembly that extend it.  Guess that works.

**Comment by namkazt** [comment]
Right. I can't even load asset bundle that was built with exactly same with Unity version of COQ. same kind of mono internal crash.

So I had another mod idea is some kind of resource loader

**Comment by namkazt** [comment]
Also i got some UI mod WIP here. plan to do some config for Harmony Injector
![2019-08-17 22_02_43-CavesOfQud](https://user-images.githubusercontent.com/6035916/63213802-d2c68400-c13a-11e9-9b2d-a8c2a5e8c161.png)

**Comment by JHawkley** [comment]
Now, that's a pretty nice one.  I pondered some kind of UI extensions myself for a short period, especially while I was figuring out how to get the popups working.

Qud's UI is a little odd in that it has two UIs running side-by-side: the console UI and the Unity Overly UI (if it is enabled).  But, it seems like a pretty frustrating situation, since you have to program and manage two different UIs at once.

Anyways, I may be wrong about that attribute thing...  'Cause rewriting to do the abstract class idea is still producing this error, and now I'm just flabbergasted.  I have no idea what is going on anymore.

I can't find any obvious errors in the code...  Nor can I find anything odd just focusing on the parts that have been diff'd.

Uh...  Wow...  So, `mcs.dll` is actually hilariously out-of-date...  Its copyright notice dates it to 2009.  Well, that explains a lot.

**Comment by namkazt** [comment]
Unity IMGUI always work without requiring to enable, so I think it is a good choice for UI

**Comment by namkazt** [comment]
This is how `Oxygen not included` restart the game.
They have additional Restarter.exe on windows
```c#
        public void Restart()
	{
		string fileName = Process.GetCurrentProcess().MainModule.FileName;
		string fullPath = Path.GetFullPath(fileName);
		string directoryName = Path.GetDirectoryName(fullPath);
		global::Debug.LogFormat("Restarting\n\texe ({0})\n\tfull ({1})\n\tdir ({2})", new object[]
		{
			fileName,
			fullPath,
			directoryName
		});
		string filename = Path.Combine(directoryName, "Restarter.exe");
		Process.Start(new ProcessStartInfo(filename)
		{
			UseShellExecute = true,
			CreateNoWindow = true,
			Arguments = string.Format("\"{0}\"", fullPath)
		});
		App.Quit();
	}

	// Token: 0x06000BAD RID: 2989 RVA: 0x000209BE File Offset: 0x0001EDBE
	public static void Quit()
	{
		Application.Quit();
	}
```


Note: In windows create 1 time scheduler is possible too.
